lib/core/util.c

```diff
@@ -64,51 +64,159 @@ static void free_accept_data(struct st_h2o_accept_data_t *data)
+{
+    h2o_redis_free((h2o_redis_conn_t *)conn);
+}
+
+h2o_redis_conn_t **redis_connection_for_context(h2o_context_t *ctx)
+{
+    static size_t key = SIZE_MAX;
+    return (h2o_redis_conn_t **)h2o_context_get_storage(ctx, &key, dispose_redis_connection);
+}
+
+static void on_redis_connect(void *data)
+{
+    fprintf(stderr, "connected to redis at %s:%" PRIu16 "\n", async_resumption_context.redis.host.base, async_resumption_context.redis.port);
+}
+
+static void on_redis_disconnect(void *data)
+{
+    fprintf(stderr, "disconnected from redis at %s:%" PRIu16 "\n", async_resumption_context.redis.host.base, async_resumption_context.redis.port);
+}
+
+static void on_redis_connection_failure(const char *errstr, void *data)
+{
+    fprintf(stderr, "redis connection failure: %s\n", errstr);
+
+    /* clear storage by setting NULL */
+    h2o_context_t *ctx = data;
+    h2o_redis_conn_t **conn = redis_connection_for_context(ctx);
+    h2o_redis_free(*conn);
+    *conn = NULL;
+}
+
+static h2o_redis_conn_t *get_redis_connection(h2o_context_t *ctx)
+{
+    h2o_redis_conn_t **conn = redis_connection_for_context(ctx);
+    if (*conn == NULL) {
+        h2o_redis_callbacks_t callbacks = (h2o_redis_callbacks_t){
+            on_redis_connect,
+            on_redis_disconnect,
+            on_redis_connection_failure,
+        };
+        *conn = h2o_redis_connect(ctx->loop, async_resumption_context.redis.host.base, async_resumption_context.redis.port, callbacks, ctx);
+    }
+
+    return *conn;
 }
 
-void h2o_accept_setup_async_ssl_resumption(h2o_memcached_context_t *memc, unsigned expiration)
+static h2o_iovec_t base64_encode(h2o_iovec_t str, h2o_iovec_t prefix)
 {
-    async_resumption_context.memc = memc;
+    h2o_iovec_t encoded;
+    encoded.base =  h2o_mem_alloc(prefix.len + (str.len + 2) / 3 * 4 + 1);
+    if (prefix.len != 0) {
+        memcpy(encoded.base, prefix.base, prefix.len);
+    }
+    encoded.len = prefix.len;
+    encoded.len += h2o_base64_encode(encoded.base + encoded.len, str.base, str.len, 1);
+    return encoded;
+}
+
+static void redis_resumption_on_get(redisReply *reply, void *_accept_data)
+{
+    struct st_h2o_accept_data_t *accept_data = _accept_data;
+
+    h2o_iovec_t session_data;
+    if (reply != NULL && reply->type == REDIS_REPLY_STRING) {
+        session_data = h2o_decode_base64url(NULL, reply->str, reply->len);
+    } else {
+        session_data = h2o_iovec_init(NULL, 0);
+    }
+
+    h2o_socket_ssl_resume_server_handshake(accept_data->sock, session_data);
+
+    if (session_data.base != NULL)
+        free(session_data.base);
+}
+
+static void redis_resumption_get(h2o_socket_t *sock, h2o_iovec_t session_id)
+{
+    struct st_h2o_accept_data_t *accept_data = sock->data;
+    h2o_redis_conn_t *conn = get_redis_connection(accept_data->ctx->ctx);
+    h2o_iovec_t key = base64_encode(session_id, async_resumption_context.redis.prefix);
+    h2o_redis_command(conn, redis_resumption_on_get, accept_data, "GET %s", key.base);
```

Comment:
> Doesn't the invocation of on_get callback get postponed until the connection is established (or times out)?
As you worried (and me too), the invocation will be postponed. I'll change the behavior like the following:
1. when the callback is called, check the connection to the redis (using `h2o_redis_is_connected` method)
2. If live connection exists, do the resumption
3. if not, async resumption fails immediately (and do full-handshake), but try to re-connect to the redis